storage/shard.go

```diff
@@ -1542,6 +1571,75 @@ func (s *dbShard) Flush(
 	return s.markFlushStateSuccessOrError(blockStart, multiErr.FinalError())
 }
 
+func (s *dbShard) Snapshot(
+	blockStart time.Time,
+	snapshotTime time.Time,
+	flush persist.Flush,
+) error {
+	// We don't snapshot data when the shard is still bootstrapping
+	s.RLock()
+	if s.bs != bootstrapped {
+		s.RUnlock()
+		return errShardNotBootstrappedToSnapshot
+	}
+	s.RUnlock()
+
+	var multiErr xerrors.MultiError
+
+	s.markIsSnapshotting()
+	defer func() {
+		s.markDoneSnapshotting(multiErr.Empty(), snapshotTime)
+	}()
+
+	prepareOpts := persist.PrepareOptions{
+		NamespaceMetadata: s.namespace,
+		Shard:             s.ID(),
+		BlockStart:        blockStart,
+		SnapshotTime:      snapshotTime,
+		FilesetType:       persist.FilesetSnapshotType,
+	}
+	prepared, err := flush.Prepare(prepareOpts)
+	// Add the err so the defer will capture it
+	multiErr = multiErr.Add(err)
+	if err != nil {
+		return err
+	}
+
+	// No action is necessary therefore we bail out early and there is no need to close.
+	if prepared.Persist == nil {
+		errMsg := "tried to write prepare snapshot file that already exists"
+		// This should never happen in practice since we don't check for duplicate snapshot files
```

Comment:
nit: Shouldn't this comment read "since we _do_ check for duplicate snapshot files in the prepare..."?

Reply:
No, the way its setup now is the Prepare method will check for duplicate Filesetfiles not snapshot files. The reason being is that snapshot files are never really duplicates, they grow incrementally and the fileset writer itself takes care of figuring out the next (auto-incrementing) index